File: rust/src/unionfind/130_solve.rs

pub fn solve(board: &mut Vec<Vec<char>>) {
    let m = board.len();
    let n = board[0].len();

    let dummy = m * n;
    
    let mut uf = UnionFind::new(dummy + 1);

    // 把board 四周的O 和dummy 相连，因为这种不会被替换
    // 首列和尾列
    for i in 0..m {
        if board[i][0] == 'O' {
            // 二维坐标(x,y) 转一维 (x*n, + y) n 是列数
            uf.union(dummy, i*n);
        }

        if board[i][n-1] == 'O' {
            uf.union(dummy, i*n + n-1);
        }
    }

    // 首行和尾行
    for i in 0..n {
        if board[0][i] == 'O' {
            uf.union(dummy,  i);
        }

        if board[m-1][i] == 'O' {
            uf.union(dummy, (m-1)*n + i);
        }
    }
    // d 是用来上下左右搜索的常用方法
    let d: Vec<Vec<i32>> = vec![vec![0,1], vec![0, -1], vec![-1,0], vec![1, 0]];
    // 遍历中间部分 遇到O 把其和四周的O 连接到一起
    for i in 1..m-1 {
        for j in 1..n-1  {
            if board[i][j] == 'O' {
                for k in 0..4 {
                    let x = (i as i32 + d[k][0]) as usize;
                    let y = (j as i32 + d[k][1]) as usize;
                    // 主要把 不是边界的O 但是和边界O 相连的过滤掉
                    if board[x][y] == 'O' {
                        uf.union(i*n+j, x*n+y);
                    }
                }
            }
        }
    }

    // 最后遍历一遍board 把不与dummy 相连的O 替换成X
    for i in 1..m {
        for j in 1..n {
            if board[i][j] == 'O' && !uf.connect(i*n+j, dummy) {
                board[i][j] = 'X';
            }
        }
    }

}

// --------------------------------------------------------------------

struct UnionFind {

    // parent[i] 表示节点i 的父节点是parent[i]
    parent: Vec<usize>,
    // 连通分量
    count: usize,
}

impl UnionFind {
    
    /* 构造函数，n 为图的节点总数 */
    fn new(n:usize) -> Self {
        // 初始时，每个节点的parent 是自己
        let mut parent = vec![0;n];
        for i in 0..n {
            parent[i] = i;
        }

        UnionFind {
            parent,
            count: n,
        }
    }


    fn find(&mut self, x: usize) -> usize {

        if self.parent[x] != x {
            // 找到x 的父节点，把x 到parent[x] 之间的所有节点压缩到parent[x] 之间
            self.parent[x] = self.find(self.parent[x]);
        }

        self.parent[x]
    }

    fn union(&mut self, from: usize, to: usize) {
        let from_root  = self.find(from);
        let to_root = self.find(to);

        if from_root == to_root {
            return;
        }
        // 找到from 和  to 的父节点，把 from 放到to 下面，或者反过来
        self.parent[from_root] = to_root;
        self.count -= 1;
    }

    fn connect(&mut self, x: usize, y:usize) -> bool {
        let root_x = self.find(x);
        let root_y = self.find(y);
        root_x == root_y
    }

    fn count(&self) -> i32 {
        self.count as i32
    }

}

File: go/unionfind/130_solve.go
package unionfind

// 但这个问题也可以用 Union-Find 算法解决，虽然实现复杂一些，甚至效率也略低，
// 但这是使用 Union-Find 算法的通用思想，值得一学。
// 你可以把那些不需要被替换的 O 看成一个拥有独门绝技的门派，
// 它们有一个共同「祖师爷」叫 dummy，这些 O 和 dummy 互相连通，而那些需要被替换的 O 与 dummy 不连通。

// 首先要解决的是，根据我们的实现，Union-Find 底层用的是一维数组，
// 构造函数需要传入这个数组的大小，而题目给的是一个二维棋盘。

// 这个很简单，二维坐标 (x,y) 可以转换成 x * n + y 这个数（m 是棋盘的行数，n 是棋盘的列数）
// 敲黑板，这是将二维坐标映射到一维的常用技巧。

// 其次，我们之前描述的「祖师爷」是虚构的，需要给他老人家留个位置。
// 索引 [0.. m*n-1] 都是棋盘内坐标的一维映射，那就让这个虚拟的 dummy 节点占据索引 m * n 好了。
func solve(board [][]byte) {
	// 1. 创建并查集
	m := len(board)
	n := len(board[0])
	// m行 n 列，转化成一维数组长度 下标 m * n -1
	// 虚拟出父节点 dummy 和dummy 相连，意味着是有连通的，不会被替换
	dummy := m * n //dummy 的坐标

	uf := NewUnionFind(dummy + 1)

	// 2. 把board 的数据放到并查集中
	// 先把首列和尾列的O 和dummy 相连，因为四边的O 不会被替换
	for i := 0; i < m; i++ {
		// 首列
		if board[i][0] == 'O' {
			// 二维转一维 (x,y) => (x*n + y) n是列数
			uf.Union(dummy, i*n)
		}
		// 尾列
		if board[i][n-1] == 'O' {
			uf.Union(dummy, i*n+n-1)
		}
	}

	// 把首行和尾行的O 和dummy 相连
	for i := 0; i < n; i++ {
		// 首行
		if board[0][i] == 'O' {
			uf.Union(dummy, i)
		}
		// 尾行
		if board[m-1][i] == 'O' {
			uf.Union(dummy, (m-1)*n+i)
		}
	}

	// 方向数组 diretion 是上下左右搜索的常用手法
	diretion := [][]int{{0, 1}, {0, -1}, {-1, 0}, {1, 0}}
	// 3. 进行Union
	for i := 1; i < m-1; i++ {
		for j := 1; j < n-1; j++ {
			// 从坐标(1,1) 开始遍历
			if board[i][j] == 'O' {
				// 将此 O 与上下左右的 O 连通
				for k := 0; k < 4; k++ {
					// x 左右移动
					x := i + diretion[k][0]
					// y 上下移动
					y := j + diretion[k][1]
					if board[x][y] == 'O' {
						// (x,y) 和 (i,j) 相连
						uf.Union(x*n+y, i*n+j)
					}
				}
			}
		}
	}
	// 4. 再遍历一次board 把不与dummy 相连的O替换成x
	for i := 1; i < m-1; i++ {
		for j := 1; j < n-1; j++ {
			if board[i][j] == 'O' && !uf.Connected(dummy, i*n+j) {
				board[i][j] = 'X'
			}

		}
	}

}

// UnionFind 并查集 用一维数组表示连通分量
type UnionFind struct {
	// parent[x] 代表 x 的父节点是parent[x]
	parent []int
	// 连通分量的数量
	count int
}

// NewUnionFind 创建长度为n 的并查集
func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	// 初始时，所有的父节点指向自己， 连通分量的数量为n
	for i := range parent {
		parent[i] = i
	}
	return &UnionFind{
		parent: parent,
		count:  n,
	}
}

// Find 找到节点x 的父节点
func (u *UnionFind) Find(x int) int {

	if u.parent[x] != x {
		// 路径压缩，把x 到根节点parent[x] 之间的所有节点，都指向parent[x]
		u.parent[x] = u.Find(u.parent[x])
	}
	return u.parent[x]
}

// Union 把节点x 和 y 连接起来
func (u *UnionFind) Union(x, y int) {
	rootX := u.Find(x)
	rootY := u.Find(y)
	if rootX == rootY {
		return
	}
	// 把x的根节点指向y的根节点，或者反过来也可以
	u.parent[rootX] = rootY
	u.count--
}

// Connected 判断x, y 是否连通，即是否有相同的父节点
func (u *UnionFind) Connected(x, y int) bool {
	rootX := u.Find(x)
	rootY := u.Find(y)

	return rootX == rootY
}

// Count 返回连通分量的个数
func (u *UnionFind) Count() int {
	return u.count
}
